src/out_ref.rs

```diff
@@ -0,0 +1,573 @@
+    ///     my_box
+    ///         .manually_drop_mut()
+    ///         .as_out()
+    /// ;
+    /// // Overwrite `my_box` with uninitialized bytes / garbage content.
+    /// unsafe {
+    ///     *at_my_box.as_uninit_mut() = MaybeUninit::uninit();
+    /// }
+    /// // Runs the destructor for a `Box<i32>` using a garbage pointer that
+    /// // may thus point anywhere in memory!
+    /// drop(my_box)
+    /// ```
+    #[inline]
+    pub
+    unsafe
+    fn as_uninit_mut (self: Out<'out, T>)
+      -> &'out mut MaybeUninit<T>
+    {
+        &mut *({self}.as_mut_ptr().cast())
+    }
+}
+
+impl<'out, T : 'out> Default for Out<'out, [T]> {
+    #[inline]
+    fn default ()
+      -> Self
+    {
+        <&mut [MaybeUninit<T>]>::into(&mut [])
+    }
+}
+
+impl<'out, T : 'out> From<&'out mut [T]> for Out<'out, [T]>
+where
+    T : Copy,
+{
+    #[inline]
+    fn from (slice: &'out mut [T])
+      -> Out<'out, [T]>
+    {
+        Out(
+            slice.into(),
+            Default::default(),
+        )
+    }
+}
+
+impl<'out, T : 'out> From<&'out mut [ManuallyDrop<T>]> for Out<'out, [T]> {
+    #[inline]
+    fn from (slice: &'out mut [ManuallyDrop<T>])
+      -> Out<'out, [T]>
+    {
+        unsafe {
+            // # Safety
+            //
+            //   - The API does not allow to write `MaybeUninit::uninit()` into
+            //     the pointee.
+            Out(
+                ptr::NonNull::new_unchecked(
+                    slice as *mut [ManuallyDrop<T>] as *mut [T]
+                ),
+                Default::default(),
+            )
+        }
+    }
+}
+
+impl<'out, T : 'out> From<&'out mut [MaybeUninit<T>]> for Out<'out, [T]> {
+    #[inline]
+    fn from (slice: &'out mut [MaybeUninit<T>])
+      -> Out<'out, [T]>
+    {
+        unsafe {
+            Out(
+                ptr::NonNull::new_unchecked(
+                    slice as *mut [MaybeUninit<T>] as *mut [T]
+                ),
+                Default::default(),
+            )
+        }
+    }
+}
+
+impl<'out, T : 'out> Out<'out, [T]> {
+    #[inline]
+    pub
+    fn as_mut_ptr (self: &'_ mut Self)
+      -> *mut T
+    {
+        self.0.as_ptr() as *mut T
+    }
+
+    /// # Safety
+    ///
+    /// The obtained reference cannot be used to write garbage
+    /// (`MaybeUninit::uninit()`) into the pointee.
+    #[inline]
+    pub
+    unsafe
+    fn as_uninit_mut (self: Out<'out, [T]>)
+      -> &'out mut [MaybeUninit<T>]
+    {
+        &mut *(self.0.as_ptr() as *mut [MaybeUninit<T>])
+    }
+
+    #[inline]
+    pub
+    fn get_out<Index> (self: Out<'out, [T]>, idx: Index)
+      -> Option<Index::Output>
+    where
+        Index : SliceIndex<'out, T>,
+    {
+        impl<'out, T : 'out> SliceIndex<'out, T> for usize {
+            type Output = Out<'out, T>;
+
+            #[inline]
+            fn idx (self: usize, slice: Out<'out, [T]>)
+              -> Option<Out<'out, T>>
+            {
+                unsafe { slice.as_uninit_mut() }
```

Comment:
Should get a `SAFETY` comment. Something about it feels strange but on first thought the implementation is actually quite clever.